### PR TITLE
WIP: New Sequence data structure (Part 2)

### PR DESCRIPTION
**Pull request by mortonjt**

### CODE REVIEW

File: skbio/io/format/genbank.py [hunk 1]
```diff
@@ -167,14 +167,13 @@
...
-    'FEATURES': <class 'list'>
     'KEYWORDS': '.'
     'LOCUS': <class 'dict'>
     'REFERENCE': <class 'list'>
     'SOURCE': <class 'dict'>
     'VERSION': '3K1V_A  GI:260656459'
 Positional metadata:
-    0: <dtype: bool>
+    <class 'skbio.sequence._feature.Feature'>: <dtype: bool>
 Stats:
```
Comment (RNAer) [inline skbio/io/format/genbank.py:177]: the repr seems problematic. there shouldn't be `Positional metadata`

**Comment by RNAer** [comment]
Add member function to combine the current interval metadata with another one. let's call it `union`? or what is the verb for `union`?

**Comment by mortonjt** [comment]
how about merge?

**Comment by mortonjt** [comment]
Added!

File: skbio/io/format/genbank.py (hunk 1 above)
Comment (mortonjt) [inline skbio/io/format/genbank.py:177]: Yeah, those doc tests are failing - will add it to the TODO list

**Comment by mortonjt** [comment]
Ok, looks like it is failing because the merge dict functionality is only available in Python 3.5

See [here](http://stackoverflow.com/questions/38987/how-can-i-merge-two-python-dictionaries-in-a-single-expression).  Will update in a bit.

**Comment by mortonjt** [comment]
1. We can update `self` instead.  But I can think of instance where it is preferable to return a copy.  What about having an `inplace` variable?
2. It will chose the features found in the later object.

``` python
In [1]: x = {1:2, 3:4}
In [2]: y = {2:2, 3:5}
In [3]: z = {**x, **y}
In [4]: z
Out[4]: {1: 2, 2: 2, 3: 5}
```

**Comment by RNAer** [comment]
:+1: for `inplace`. The update with `other` is fine, but should be clearly
documented - put `..warning: XXXX` in the docstring.

**Comment by RNAer** [comment]
ok. I'll merge this for now. don't forget fix the doctest later.

File: skbio/io/format/genbank.py [hunk 2]
```diff
@@ -709,6 +730,63 @@ def _serialize_qualifier(key, value):
...
+            index = []
+        else:
+            raise GenBankFormatError(
+                'Could not parse location string: "%s"' %
+                loc_str)
+        intervals += index
+
+    return res, intervals
 
 def _parse_loc_str(loc_str, length):
```
Comment (RNAer) [inline skbio/io/format/genbank.py:791]: this func seems replaced by _parse_interval? delete it?

Comment (mortonjt) [inline skbio/io/format/genbank.py:791]: Done

**Comment by mortonjt** [comment]
Hi @jairideout @ebolyen,  I'm discussing with @RNAer about where is the best place to organize these files.

Right now, all of the interval metadata objects are scattered under `skbio.sequence`.
However, I'm thinking that it would be a better idea to lump the `IntervalMetadataMixin` object in `_base.py` for a few reasons
1. This object will be organized with the other metadata objects
2. It will be more straightforward for organizing other metadata methods i.e. `_MetadataReprBuilder`
3. It'll circumvent some circular import issues.

So what your thoughts?  Should we also keep `_interval.py`, `_feature.py` in the `sequence` folder, or would they be more appropriate outside (i.e. in `util` or the root directory?).

**Comment by jairideout** [comment]
I don't have strong feelings about where this functionality lives because it's private (we can easily move it without affecting users). It probably makes sense to put the interval metadata functionality where we have the other metadata mixins.

**Comment by RNAer** [comment]
@jairideout, not all are private. the `IntervalMetadata` class should be
public

**Comment by jairideout** [comment]
How about putting all metadata-related functionality under a new `skbio.metadata` subpackage? `skbio._base` is getting cluttered, so it'd be nice to move the metadata mixins there too.

**Comment by ebolyen** [comment]
:+1: I like that.

**Comment by mortonjt** [comment]
Awesome!

Just to make sure that we're on the same page. In the context of these PRs, the skbio.metadata subpackage will include the following: 
- All of the MetadataMixin objects
- `_metadata_repr.py`
- `_interval.py` - the interval_metadata object
- `_feature.py` - the feature hashable object
- `intersection.pyx` - the actual IntervalTree datastructure.

Is this inline what you were thinking @jairideout ?

**Comment by jairideout** [comment]
Yep! Does it make sense to have all public functionality importable from `skbio.metadata`?

**Comment by mortonjt** [comment]
For the `IntervalMetadata` object and the `Feature` object, I think it makes sense.  I could see embedding Features within `skbio.TreeNode`.

However, I'm not sure about IntervalTree ...

**Comment by ebolyen** [comment]
I don't think `IntervalTree` should be public, it's a detail of `IntervalMetadata`.

**Comment by jairideout** [comment]
:+1:

**Pull request merged**